src/core/GizmoHelper.tsx

```diff
@@ -0,0 +1,121 @@
+}
+
+const Context = React.createContext<GizmoHelperContext>({} as GizmoHelperContext)
+
+export const useGizmoHelper = () => {
+  return React.useContext<GizmoHelperContext>(Context)
+}
+
+const turnRate = 2 * Math.PI // turn rate in angles per second
+const dummy = new Object3D()
+const matrix = new Matrix4()
+const [q1, q2] = [new Quaternion(), new Quaternion()]
+const target = new Vector3()
+const targetPosition = new Vector3()
+const targetQuaternion = new Quaternion()
+
+export type GizmoHelperProps = JSX.IntrinsicElements['group'] & {
+  alignment?: 'top-left' | 'top-right' | 'bottom-right' | 'bottom-left'
+  margin?: [number, number]
+  onUpdate: () => void // update controls during animation
+  onTarget: () => Vector3 // return the target to rotate around
+}
+
+export const GizmoHelper = ({
+  alignment = 'bottom-right',
+  margin = [80, 80],
+  onUpdate,
+  onTarget,
+  children: GizmoHelperComponent,
+}: GizmoHelperProps): any => {
+  const { gl, scene, camera: mainCamera, size } = useThree()
+  const gizmoRef = React.useRef<Group>()
+  const virtualCam = React.useRef<any>()
+  const virtualScene = React.useMemo(() => new Scene(), [])
+  const [animating, setAnimating] = React.useState(false)
+  const [radius, setRadius] = React.useState(0)
+  const [focusPoint, setFocusPoint] = React.useState(new Vector3(0, 0, 0))
+
+  function tweenCamera(direction: Vector3) {
+    const radius = mainCamera.position.distanceTo(target)
+    setRadius(radius)
+    setFocusPoint(onTarget())
+    setAnimating(true)
+
+    // Rotate from current camera orientation
+    dummy.position.copy(target)
+    dummy.lookAt(mainCamera.position)
+    q1.copy(dummy.quaternion)
+
+    // To new current camera orientation
+    targetPosition.copy(direction).multiplyScalar(radius).add(target)
+    dummy.lookAt(targetPosition)
+    q2.copy(dummy.quaternion)
+  }
+
+  function animateStep(delta: number): void {
+    if (!animating) return
+
+    const step = delta * turnRate
+
+    // animate position by doing a slerp and then scaling the position on the unit sphere
+    q1.rotateTowards(q2, step)
+    mainCamera.position.set(0, 0, 1).applyQuaternion(q1).multiplyScalar(radius).add(focusPoint)
+
+    // animate orientation
+    mainCamera.quaternion.rotateTowards(targetQuaternion, step)
+    mainCamera.updateProjectionMatrix()
+    onUpdate && onUpdate()
+
+    if (q1.angleTo(q2) === 0) {
+      setAnimating(false)
+    }
+  }
+
+  useFrame((_, delta) => {
+    if (virtualCam.current && gizmoRef.current) {
+      animateStep(delta)
+
+      // Sync gizmo with main camera orientation
+      matrix.copy(mainCamera.matrix).invert()
+      gizmoRef.current.quaternion.setFromRotationMatrix(matrix)
+
+      // Render main scene
+      gl.autoClear = true
+      gl.render(scene, mainCamera)
+
+      // Render gizmo
+      gl.autoClear = false
+      gl.clearDepth()
+      gl.render(virtualScene, virtualCam.current)
+    }
+  }, 1)
```

Comment:
We've tried not to hijack the rendering order in our components, can you pleas elaborate why you feel the need to?

Reply:
I don't think it's needed and I've removed the priority. How do I ensure that the gizmo is always rendered on top of the main scene?

Reply:
*maybe* this – https://github.com/pmndrs/react-three-fiber/blob/master/markdown/recipes.md#heads-up-display-rendering-multiple-scenes ?

Reply:
I'm a little stuck here. Is there a working code example for the recipe you linked? I couldn't make it work and don't see any examples that follow the recipe. 

Both https://codesandbox.io/s/react-three-fiber-viewcube-py4db?file=/src/App.js and https://github.com/pmndrs/drei/blob/master/.storybook/stories/useCamera.stories.tsx render the two scenes in one useFrame call similar to the GizmoHelper code.

Reply:
This forked example removes the render-order priority - https://codesandbox.io/s/react-three-fiber-viewcube-forked-4zu7d and appears to still render the cube on top of the scene. I can look at the code later it this isn't the case for your component.

Reply:
@joshuaellis thank you for the sandbox! Seems to be working well now :)